server/game/gameSystems/DiscardFromDeckSystem.ts

```diff
@@ -0,0 +1,89 @@
+import Player from '../core/Player';
+import { DiscardSpecificCardSystem } from './DiscardSpecificCardSystem';
+import * as Contract from '../core/utils/Contract';
+
+export interface IDiscardFromDeckProperties extends IPlayerTargetSystemProperties {
+    amount?: number;
+}
+
+export class DiscardFromDeckSystem<TContext extends AbilityContext = AbilityContext> extends PlayerTargetSystem<TContext, IDiscardFromDeckProperties> {
+    public override readonly name = 'discardFromDeck';
+    public override readonly eventName = EventName.OnDiscardFromDeck;
+
+    protected override defaultProperties: IDiscardFromDeckProperties = {
+        amount: 1
+    };
+
+    // eslint-disable-next-line @typescript-eslint/no-empty-function
+    public override eventHandler(_event): void { }
+
+    // public eventHandler(event): void {
+    //     for (let i = 0; i < event.amount; i++) {
+    //         const topCard = event.player.getTopCardOfDeck();
+    //         event.player.moveCard(topCard, Location.Discard);
+    //     }
+    // }
+
+    public override getEffectMessage(context: TContext): [string, any[]] {
+        const properties = this.generatePropertiesFromContext(context);
+        return ['discard ' + (properties.amount + (properties.amount > 1 ? ' cards' : ' card') + ' from deck'), []];
+    }
+
+    public override canAffect(player: Player, context: TContext, additionalProperties = {}, mustChangeGameState = GameStateChangeRequired.None): boolean {
+        const properties = this.generatePropertiesFromContext(context, additionalProperties);
+
+        const availableDeck = player.drawDeck || []; // Without this | [], sometimes this is undefined, causing issues below
+
+        if (mustChangeGameState !== GameStateChangeRequired.None && (availableDeck.length === 0 || properties.amount === 0)) {
+            return false;
+        }
+
+        if ((properties.isCost || GameStateChangeRequired.MustFullyResolve) && availableDeck.length < properties.amount) {
+            return false;
+        }
+
+        if (!super.canAffect(player, context, additionalProperties)) {
+            return false;
+        }
+
+        return properties.amount !== 0 && super.canAffect(player, context);
+    }
+
+    protected override addPropertiesToEvent(event, player: Player, context: TContext, additionalProperties): void {
+        const { amount } = this.generatePropertiesFromContext(context, additionalProperties);
+        super.addPropertiesToEvent(event, player, context, additionalProperties);
+        event.amount = amount;
+    }
+
+    public override queueGenerateEventGameSteps(events: any[], context: TContext, additionalProperties: Record<string, any> = {}): void {
+        const properties = this.generatePropertiesFromContext(context, additionalProperties);
+        for (const player of properties.target as Player[]) {
+            const availableDeck = player.drawDeck;
+
+            Contract.assertNonNegative(properties.amount);
```

Comment:
Actually sorry now that I think about it, could we move this to `canAffect` since that gets hit earlier and more often